go/tools/gazelle/merger/fix.go

```diff
@@ -0,0 +1,189 @@
+/* Copyright 2017 The Bazel Authors. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package merger
+
+import (
+	"sort"
+
+	bf "github.com/bazelbuild/buildtools/build"
+)
+
+// FixFile updates rules in oldFile that were generated by an older version of
+// Gazelle to a newer form that can be merged with freshly generated rules.
+func FixFile(oldFile *bf.File) *bf.File {
+	return fixLoads(oldFile)
+}
+
+// fixLoads removes loads of unused go rules and adds loads of newly used rules.
+func fixLoads(oldFile *bf.File) *bf.File {
+	// Identify load statements for the Go rules, and determine which symbols
+	// are actually used.
+	type loadInfo struct {
+		index      int
+		old, fixed *bf.CallExpr
+	}
+	var loads []loadInfo
+	usedKinds := make(map[string]bool)
+	for i, stmt := range oldFile.Stmt {
+		c, ok := stmt.(*bf.CallExpr)
+		if !ok {
+			continue
+		}
+		x, ok := c.X.(*bf.LiteralExpr)
+		if !ok {
+			continue
+		}
+
+		if x.Token == "load" {
+			if len(c.List) == 0 {
+				continue
+			}
+			if label, ok := c.List[0].(*bf.StringExpr); ok && label.Value == "@io_bazel_rules_go//go:def.bzl" {
```

Comment:
We have talked about the fact that io_bazel_rules_go is the wrong name, but it's too hard to change.
Should we extract it to a constant, and think about wether the new fixing ability could update this name for us?